### PR TITLE
add test for gc raft log

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
@BusyJay @siddontang PTAL

### CODE REVIEW

File: src/raftstore/store/worker/raftlog_gc.rs [hunk 1]
```diff
@@ -51,9 +56,15 @@ quick_error! {
     }
 }
 
-pub struct Runner;
+pub struct Runner {
+    ch: Option<Sender<TaskRes>>,
```
Comment (siddontang) [inline src/raftstore/store/worker/raftlog_gc.rs]: I think we can send the task, wait a little time and then check the rocksdb data directly, not the sender.

Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/raftlog_gc.rs]: I think use channel and receive the `TaskRes` is more clear.

File: src/raftstore/store/worker/raftlog_gc.rs [hunk 2]
```diff
@@ -79,10 +90,17 @@ impl Runner {
...
+        // It's not safe to disable WAL here.
         engine.write(wb).unwrap();
         Ok(end_idx - first_idx)
     }
+
+    fn finish_task(&mut self, collected: u64) {
+        if self.ch.is_none() {
+            return;
+        }
+        self.ch.as_mut().unwrap().send(TaskRes { collected: collected }).unwrap();
```
Comment (siddontang) [inline src/raftstore/store/worker/raftlog_gc.rs]: no need mutable for channel send

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/raftstore/store/worker/raftlog_gc.rs [hunk 3]
```diff
@@ -79,10 +90,17 @@ impl Runner {
             let key = keys::raft_log_key(region_id, idx);
             box_try!(wb.delete_cf(handle, &key));
         }
-        // It's not safe to disable WAL here. We may lost data after crashed for unknown reason.
+        // It's not safe to disable WAL here.
```
Comment (BusyJay) [inline src/raftstore/store/worker/raftlog_gc.rs]: Why it's not safe to disable WAL here?

File: src/raftstore/store/worker/raftlog_gc.rs [hunk 4]
```diff
@@ -79,10 +90,17 @@ impl Runner {
...
         engine.write(wb).unwrap();
         Ok(end_idx - first_idx)
     }
+
+    fn finish_task(&self, collected: u64) {
```
Comment (BusyJay) [inline src/raftstore/store/worker/raftlog_gc.rs]: s/finish_task/report_collected/

File: src/raftstore/store/worker/raftlog_gc.rs [hunk 5]
```diff
@@ -91,8 +109,84 @@ impl Runnable<Task> for Runner {
...
+
+        // gc 0..10
+        runner.run(Task {
+            engine: db.clone(),
+            region_id: region_id,
+            start_idx: 0,
+            end_idx: 10,
+        });
+        let res = rx.recv_timeout(Duration::from_secs(3)).unwrap();
+        assert_eq!(res.collected, 10);
```
Comment (BusyJay) [inline src/raftstore/store/worker/raftlog_gc.rs]: Should also check if the log is removed.

File: src/raftstore/store/worker/raftlog_gc.rs [hunk 6]
```diff
@@ -188,5 +194,23 @@ mod test {
...
         assert_eq!(res.collected, 10);
+        raft_log_must_not_exist(&db, 1, 0, 60);
+        raft_log_must_exist(&db, 1, 60, 100);
+    }
+
+    fn raft_log_must_not_exist(engine: &DB, region_id: u64, start_idx: u64, end_idx: u64) {
+        let raft_handle = rocksdb::get_cf_handle(engine, CF_RAFT).unwrap();
+        for i in start_idx..end_idx {
+            let k = keys::raft_log_key(region_id, i);
+            engine.get_cf(raft_handle, &k).unwrap().is_none();
```
Comment (BusyJay) [inline src/raftstore/store/worker/raftlog_gc.rs]: Should assert.

File: src/raftstore/store/worker/raftlog_gc.rs [hunk 7]
```diff
@@ -79,10 +90,17 @@ impl Runner {
             let key = keys::raft_log_key(region_id, idx);
             box_try!(wb.delete_cf(handle, &key));
         }
-        // It's not safe to disable WAL here. We may lost data after crashed for unknown reason.
+        // TODO: disable WAL here.
```
Comment (BusyJay) [inline src/raftstore/store/worker/raftlog_gc.rs]: /cc @siddontang Is is safe to disable WAL here?

Comment (siddontang) [inline src/raftstore/store/worker/raftlog_gc.rs]: Seem that RocksDB has already fixed this bug.

Comment (siddontang) [inline src/raftstore/store/worker/raftlog_gc.rs]: But we should do more tests later to verify this.

File: src/raftstore/store/worker/raftlog_gc.rs [hunk 8]
```diff
@@ -91,8 +109,108 @@ impl Runnable<Task> for Runner {
...
+        let region_id = 1;
+        let wb = WriteBatch::new();
+        for i in 0..100 {
+            let k = keys::raft_log_key(region_id, i);
+            wb.put_cf(raft_handle, &k, b"entry").unwrap();
+        }
+        db.write(wb).unwrap();
+
+        // gc 0..10
+        runner.run(Task {
```
Comment (siddontang) [inline src/raftstore/store/worker/raftlog_gc.rs]:
```
let tbls = vec![(task, collected, (not exist range), (exist range)),]
for tbl in tbls {
    run_task
    check collected
    check not exist
    check exist
}
```

Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/raftlog_gc.rs]: I think it is more clear that unroll the short loop in test.

Comment (siddontang) [inline src/raftstore/store/worker/raftlog_gc.rs]: If we can use a table to drive the test, we should use it.

File: src/raftstore/store/worker/raftlog_gc.rs [hunk 9]
```diff
@@ -91,8 +109,109 @@ impl Runnable<Task> for Runner {
...
+                            engine: db.clone(),
+                            region_id: region_id,
+                            start_idx: 50,
+                            end_idx: 60,
+                        },
+                         10,
+                         (0, 60),
+                         (60, 100))];
+
+        for tbl in tbls {
```
Comment (siddontang) [inline src/raftstore/store/worker/raftlog_gc.rs]:
```
for (task, collected, not_exist_range, exist_range) in tbl {

}
```

**Comment by siddontang** [comment]
Rest LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**